crates/loader/src/validation.rs

```diff
@@ -1,23 +1,276 @@
 
-// Check whether http host can be parsed by Url
+// Checks a list of allowed HTTP hosts is valid
 pub fn validate_allowed_http_hosts(http_hosts: &Option<Vec<String>>) -> Result<()> {
-    if let Some(domains) = http_hosts.as_deref() {
-        if domains
-            .iter()
-            .any(|domain| domain == wasi_outbound_http::ALLOW_ALL_HOSTS)
-        {
-            return Ok(());
+    parse_allowed_http_hosts(http_hosts).map(|_| ())
+}
+
+// Parses a list of allowed HTTP hosts
+pub fn parse_allowed_http_hosts(raw: &Option<Vec<String>>) -> Result<AllowedHttpHosts> {
+    match raw {
+        None => Ok(AllowedHttpHosts::AllowSpecific(vec![])),
+        Some(list) => {
+            if list
+                .iter()
+                .any(|domain| domain == wasi_outbound_http::ALLOW_ALL_HOSTS)
+            {
+                Ok(AllowedHttpHosts::AllowAll)
+            } else {
+                let parse_results = list
+                    .iter()
+                    .map(|h| parse_allowed_http_host(h))
+                    .collect::<Vec<_>>();
+                let (hosts, errors) = partition_results(parse_results);
+
+                if errors.is_empty() {
+                    Ok(AllowedHttpHosts::AllowSpecific(hosts))
+                } else {
+                    Err(anyhow!(
+                        "One or more allowed_http_hosts entries was invalid:\n{}",
+                        errors.join("\n")
+                    ))
+                }
+            }
+        }
+    }
+}
+
+fn parse_allowed_http_host(text: &str) -> Result<AllowedHttpHost, String> {
+    // If you call Url::parse, it accepts things like `localhost:3001`, inferring
+    // `localhost` as a scheme. That's unhelpful for us, so we do a crude check
+    // before trying to treat the string as a URL.
+    if text.contains("//") {
+        parse_allowed_http_host_as_schemed_url(text)
+    } else {
+        parse_allowed_http_host_as_host(text)
+    }
+}
+
+fn parse_allowed_http_host_as_host(text: &str) -> Result<AllowedHttpHost, String> {
+    match text.split(':').collect::<Vec<_>>()[..] {
+        [host] => match is_valid_host_name(host) {
+            true => Ok(AllowedHttpHost::host(host)),
+            false => Err(format!("{} isn't a valid host or host:port string", text)),
+        },
+        [host, port] => match (is_valid_host_name(host), u16::from_str(port)) {
+            (true, Ok(p)) => {
+                if is_allowed_port(p) {
+                    Ok(AllowedHttpHost::host_and_port(host, p))
+                } else {
+                    Err(format!(
+                        "Can't allow {} because the port is reserved for non-HTTP protocols",
+                        text
+                    ))
+                }
+            }
+            _ => Err(format!("{} isn't a valid host or host:port string", text)),
+        },
+        _ => Err(format!("{} isn't a valid host or host:port string", text)),
+    }
+}
+
+fn is_valid_host_name(text: &str) -> bool {
```

Comment:
Is this check necessary? It will reject IPv6 addrs as-is.

Reply:
Hmm, and the way I'm checking for the presence of a port will also misfire on IPv6 addresses.

_mutters darkly_